Lib/importlib/_common.py

```diff
@@ -0,0 +1,72 @@
+import os
+import pathlib
+import zipfile
+import tempfile
+import functools
+import contextlib
+
+
+def from_package(package):
+    """
+    Return a Traversable object for the given package.
+
+    """
+    spec = package.__spec__
+    return from_traversable_resources(spec) or fallback_resources(spec)
+
+
+def from_traversable_resources(spec):
+    """
+    If the spec.loader implements TraversableResources,
+    directly or implicitly, it will have a ``files()`` method.
+    """
+    with contextlib.suppress(AttributeError):
+        return spec.loader.files()
+
+
+def fallback_resources(spec):
+    package_directory = pathlib.Path(spec.origin).parent
+    try:
+        archive_path = spec.loader.archive
+        rel_path = package_directory.relative_to(archive_path)
+        return zipfile.Path(archive_path, str(rel_path) + '/')
+    except Exception:
+        pass
+    return package_directory
```

Comment:
From my understanding, I believe only `ValueError` and `OSError` should be reasonably expected to occur within the try block here. If I'm not missing something, I think this could be changed to:

```suggestion
def fallback_resources(spec):
    package_directory = pathlib.Path(spec.origin).parent
    try:
        archive_path = spec.loader.archive
        rel_path = package_directory.relative_to(archive_path)
        return zipfile.Path(archive_path, str(rel_path) + '/')
    except (ValueError, OSError):
        pass
    return package_directory
```

Reply:
I tried that and ran the tests and the test suite triggered AttributeError on `spec.loader` (no `archive`). Agreed it would be nice to capture all of the minimum envelope necessary, but at the point that we're listing three exceptions, I'm inclined to just capture all of them.

Reply:
Given that this is the implementation in importlib_resources, I'd like to stick with it for now, and I'd like to get this merged for the last alpha to get some early feedback. I expect lots of tuning tweaks to the implementation as we work through the betas.

Reply:
Okay, it's probably fine then. If we'd have to explicitly capture 3 exception classes (if not more, as I'm not aware of all the exceptions that can potentially be raised from `spec.loader`), it doesn't really make much of a difference from `except Exception`.